### PR TITLE
Fix HDFS with AD auth for browse, read file scenarios

### PR DESCRIPTION
**Pull request by kevcunnane**
- HDFS now fully supports expanding nodes for all levels, including using cookie for auth
- HDFS now support reading files from HDFS
- HDFS write file is broken and will be fixed (either in PR update or separate PR)
- Removed hack to use gateway-0 instead of actual DNS name now these are supported. Needed for testing

### CODE REVIEW

File: extensions/mssql/package.json [hunk 1]
```diff
@@ -949,7 +931,29 @@
...
+    "error-ex": "^1.3.2",
+    "figures": "^2.0.0",
+    "find-remove": "1.2.1",
+    "fs-extra": "^3.0.1",
+    "kerberos": "^1.1.2",
+    "request": "^2.88.0",
+    "request-promise": "^4.2.2",
+    "service-downloader": "github:anthonydresser/service-downloader#0.1.6",
+    "stream-meter": "^1.0.4",
+    "through2": "^3.0.1",
```
Comment (kevcunnane) [inline extensions/mssql/package.json]: Added through-2 and tought-cookie in this change. Both were already in yarn.lock via transitive dependency chaining (e.g. we already had these, I just made it official here)

File: extensions/mssql/src/objectExplorerNodeProvider/webhdfs.ts [hunk 2]
```diff
@@ -6,18 +6,20 @@ import * as fs from 'fs';
...
+import { Cookie } from 'tough-cookie';
+import * as through from 'through2';
 import * as nls from 'vscode-nls';
+import * as auth from '../util/auth';
 import { IHdfsOptions, IRequestParams } from './fileSources';
 
 const localize = nls.loadMessageBundle();
-const ErrorMessageInvalidDataStructure =
-	localize('webhdfs.invalidDataStructure', 'Invalid Data Structure');
+const ErrorMessageInvalidDataStructure = localize('webhdfs.invalidDataStructure', 'Invalid Data Structure');
```
Comment (alanrenmsft) [inline extensions/mssql/src/objectExplorerNodeProvider/webhdfs.ts]: double quote the actual string

File: extensions/mssql/src/objectExplorerNodeProvider/webhdfs.ts [hunk 3]
```diff
@@ -575,21 +645,51 @@ export class WebHDFS {
...
+		let params: request.OptionsWithUrl = Object.assign(
 			{
 				method: 'GET',
 				url: endpoint,
 				json: true
 			},
 			this._requestParams
 		);
+		if (!this._requestParams.isKerberos) {
+			return <fs.ReadStream><any>this.doCreateReadStream(params);
```
Comment (alanrenmsft) [inline extensions/mssql/src/objectExplorerNodeProvider/webhdfs.ts]: what does <fs.ReadStream><any> do here?

Comment (kevcunnane) [reply extensions/mssql/src/objectExplorerNodeProvider/webhdfs.ts]: It just lets upstream callers treat this as a ReadStream so they know what events to expect via the pipe

**Pull request merged**